tutorials/src/bin/basic-tutorial-5.rs

```diff
@@ -0,0 +1,362 @@
+                            &format!("    bitrate: {} \n", bitrate.get().unwrap()),
+                        );
+                    }
+                }
+            },
+            None => {
+                eprintln!("Could not get {}!", propname);
+            }
+        }
+    }
+
+    // Extract metadata from all the streams and write it to the text widget in the GUI
+    fn analyze_streams(playbin: &gst::Element, textbufcell: &SendCell<gtk::TextBuffer>) {
+        {
+            let textbuf = textbufcell.borrow();
+            textbuf.set_text(&"");
+        }
+
+        add_streams_info(playbin, textbufcell, "video");
+        add_streams_info(playbin, textbufcell, "audio");
+        add_streams_info(playbin, textbufcell, "text");
+    }
+
+    // This creates all the GTK+ widgets that compose our application, and registers the callbacks
+    pub fn create_ui(playbin: &gst::Element) {
+        let main_window = Window::new(WindowType::Toplevel);
+        main_window.connect_delete_event(|_, _| {
+            gtk::main_quit();
+            Inhibit(false)
+        });
+
+        let play_button = gtk::Button::new_from_icon_name(
+            "media-playback-start",
+            gtk::IconSize::SmallToolbar.into(),
+        );
+        let pipeline = playbin.clone();
+        play_button.connect_clicked(move |_| {
+            let pipeline = &pipeline;
+            pipeline.set_state(gst::State::Playing);
+        });
+
+        let pause_button = gtk::Button::new_from_icon_name(
+            "media-playback-pause",
+            gtk::IconSize::SmallToolbar.into(),
+        );
+        let pipeline = playbin.clone();
+        pause_button.connect_clicked(move |_| {
+            let pipeline = &pipeline;
+            pipeline.set_state(gst::State::Paused);
+        });
+
+        let stop_button = gtk::Button::new_from_icon_name(
+            "media-playback-stop",
+            gtk::IconSize::SmallToolbar.into(),
+        );
+        let pipeline = playbin.clone();
+        stop_button.connect_clicked(move |_| {
+            let pipeline = &pipeline;
+            pipeline.set_state(gst::State::Ready);
+        });
+
+        let slider = gtk::Scale::new_with_range(
+            gtk::Orientation::Horizontal,
+            0.0 as f64,
+            100.0 as f64,
+            1.0 as f64,
+        );
+        let pipeline = playbin.clone();
+        let slider_update_signal_id = slider
+            .connect_value_changed(move |slider| {
+                let pipeline = &pipeline;
+                let value = slider.get_value() as u64;
+                if let Err(_) = pipeline .seek_simple(
+                    gst::Format::Time,
+                    gst::SeekFlags::FLUSH | gst::SeekFlags::KEY_UNIT,
+                    ((value * gst::SECOND) as i64),
+                    )
+                {
+                    eprintln!("Seeking to {} failed", value);
+                }
+            })
+            .to_glib();
```

Comment:
to_glib() also shouldn't be here. The from_glib_* and to_glib_* functions are all meant for the bindings code only, or when you interact with the C functions directly